Tests/NIOTests/ByteBufferTest.swift

```diff
@@ -1214,6 +1214,87 @@ class ByteBufferTest: XCTestCase {
     }
+
+    func testByteBufferFitsInACoupleOfEnums() throws {
+        enum Level4 {
+            case case1(ByteBuffer)
+            case case2(ByteBuffer)
+            case case3(ByteBuffer)
+            case case4(ByteBuffer)
+        }
+        enum Level3 {
+            case case1(Level4)
+            case case2(Level4)
+            case case3(Level4)
+            case case4(Level4)
+        }
+        enum Level2 {
+            case case1(Level3)
+            case case2(Level3)
+            case case3(Level3)
+            case case4(Level3)
+        }
+        enum Level1 {
+            case case1(Level2)
+            case case2(Level2)
+            case case3(Level2)
+            case case4(Level2)
+        }
+
+        XCTAssertLessThanOrEqual(MemoryLayout<ByteBuffer>.size, 23)
+        XCTAssertLessThanOrEqual(MemoryLayout<Level1>.size, 24)
+
+        XCTAssertLessThanOrEqual(MemoryLayout.size(ofValue: Level1.case1(.case2(.case3(.case4(self.buf))))), 24)
+        XCTAssertLessThanOrEqual(MemoryLayout.size(ofValue: Level1.case1(.case3(.case4(.case1(self.buf))))), 24)
+    }
+
+    func testLargeSliceBegin16MBIsOkayAndDoesNotCopy() throws {
+        var fourMBBuf = self.allocator.buffer(capacity: 4 * 1024 * 1024)
+        fourMBBuf.write(bytes: repeatElement(0xff, count: fourMBBuf.capacity))
+        let totalBufferSize = 5 * fourMBBuf.readableBytes
+        XCTAssertEqual(4 * 1024 * 1024, fourMBBuf.readableBytes)
+        var buf = self.allocator.buffer(capacity: totalBufferSize)
+        for _ in 0..<5 {
+            var fresh = fourMBBuf
+            buf.write(buffer: &fresh)
+        }
+        XCTAssertEqual(totalBufferSize, buf.readableBytes)
+
+        let offset = Int(_UInt24.max)
+
+        let oldPtrVal = buf.withUnsafeReadableBytes {
+            UInt(bitPattern: $0.baseAddress!.advanced(by: offset))
+        }
+
+        let slice = buf.getSlice(at: offset, length: 4 * 1024 * 1024)!
+        XCTAssertEqual(4 * 1024 * 1024, slice.readableBytes)
+        let newPtrVal = slice.withUnsafeReadableBytes {
+            UInt(bitPattern: $0.baseAddress!)
+        }
+        XCTAssertEqual(oldPtrVal, newPtrVal)
+    }
+
+    func testLargeSliceBeginMoreThan16MBIsOkay() throws {
+        var fourMBBuf = self.allocator.buffer(capacity: 4 * 1024 * 1024)
+        fourMBBuf.write(bytes: repeatElement(0xff, count: fourMBBuf.capacity))
+        let totalBufferSize = 5 * fourMBBuf.readableBytes
+        XCTAssertEqual(4 * 1024 * 1024, fourMBBuf.readableBytes)
+        var buf = self.allocator.buffer(capacity: totalBufferSize)
+        for _ in 0..<5 {
+            var fresh = fourMBBuf
+            buf.write(buffer: &fresh)
+        }
+        XCTAssertEqual(totalBufferSize, buf.readableBytes)
+
+        let offset = Int(_UInt24.max) + 1
+        let expectedReadablebytes = totalBufferSize - offset
+        let slice = buf.getSlice(at: offset, length: expectedReadablebytes)!
```

Comment:
Minor nit: it might be better if we sliced the same size as the test above, to isolate the test differences to only being the slice start index, rather than the slice sizes.

Reply:
addressed and improved test (by marking some 'important' positions with different byte values)